openquake/baselib/general.py

```diff
@@ -218,100 +218,40 @@ def split_in_blocks(sequence, hint, weight=lambda item: 1,
+    """
+    assert len(seq1) == len(seq2), 'Lists of different lenghts: %d != %d' % (
+        len(seq1), len(seq2))
+    for x, y in zip(seq1, seq2):
+        assert_close(x, y, rtol, atol)
 
 
-def _deep_eq(a, b, decimal, exclude=None):
-    """Do the actual deep comparison. If the two items up for comparison are
-    not equal, a :exception:`AssertionError` is raised (to
-    :function:`deep_eq`).
+def assert_close(a, b, rtol=1e-07, atol=0):
     """
-
-    exclude = exclude or []
-
-    def _test_dict(a, b):
-        """Compare `dict` types recursively."""
-        assert len(a) == len(b), (
-            "Dicts %(a)s and %(b)s do not have the same length."
-            " Actual lengths: %(len_a)s and %(len_b)s") % dict(
-            a=a, b=b, len_a=len(a), len_b=len(b))
-
-        for key in a:
-            if not key in exclude:
-                _deep_eq(a[key], b[key], decimal)
-
-    def _test_seq(a, b):
-        """Compare `list` or `tuple` types recursively."""
-        assert len(a) == len(b), (
-            "Sequences %(a)s and %(b)s do not have the same length."
-            " Actual lengths: %(len_a)s and %(len_b)s") % \
-            dict(a=a, b=b, len_a=len(a), len_b=len(b))
-
-        for i, item in enumerate(a):
-            _deep_eq(item, b[i], decimal)
-
-    # lists or tuples
-    if isinstance(a, (list, tuple)):
-        _test_seq(a, b)
-    # dicts
-    elif isinstance(a, dict):
-        _test_dict(a, b)
-    # objects with a __dict__
-    elif hasattr(a, '__dict__'):
-        assert a.__class__ == b.__class__, (
-            "%s and %s are different classes") % (a.__class__, b.__class__)
-        _test_dict(a.__dict__, b.__dict__)
-    elif isinstance(a, numpy.ndarray):
-        assert numpy.array_equal(a, b), '%s and %s are different' % (a, b)
-    # iterables (not strings)
-    elif isinstance(a, collections.Iterable) and not isinstance(a, str):
-        # If there's a generator or another type of iterable, treat it as a
-        # `list`. NOTE: Generators will be exhausted if you do this.
-        _test_seq(list(a), list(b))
-    # objects with __slots__
-    elif hasattr(a, '__slots__'):
-        assert a.__class__ == b.__class__, (
-            "%s and %s are different classes") % (a.__class__, b.__class__)
-        assert a.__slots__ == b.__slots__, (
-            "slots %s and %s are not the same") % (a.__slots__, b.__slots__)
-        for slot in a.__slots__:
-            if slot not in exclude:
-                _deep_eq(getattr(a, slot), getattr(b, slot), decimal)
-    else:
-        # Objects must be primitives
-
-        # Are they numbers?
-        if isinstance(a, (int, long, float, complex)):
-            numpy.testing.assert_almost_equal(a, b, decimal=decimal)
-        else:
-            assert a == b, "%s != %s" % (a, b)
+    Compare for equality up to a given precision two composite objects
+    which may contain floats. NB: if the objects are or contain generators,
+    they are exhausted.
```

Comment:
Yes, but it would complicate the implementation too much. There is no good way to copy iterators in Python (itertools.tee is no good).